### PR TITLE
RHINENG-8141: improve evaluation performance

### PR DESCRIPTION
**Pull request by psegedy**
## Secure Coding Practices Checklist GitHub Link
- https://github.com/RedHatInsights/secure-coding-checklist

## Secure Coding Checklist
- [x] Input Validation
- [x] Output Encoding
- [x] Authentication and Password Management
- [x] Session Management
- [x] Access Control
- [x] Cryptographic Practices
- [x] Error Handling and Logging
- [x] Data Protection
- [x] Communication Security
- [x] System Configuration
- [x] Database Security
- [x] File Management
- [x] Memory Management
- [x] General Coding Practices

### CODE REVIEW

**Review by psegedy — COMMENTED** [review]
@Dugowitch sorry for my mistake with the PR, please open a new one from your fork once you introduce changes. Overall, the logic looks good, there are only some minor things I want you to look at.

File: evaluator/evaluate_advisories.go [hunk 1]
```diff
@@ -12,162 +12,146 @@ import (
...
+		}
 	}
-	installableIDs = append(installableIDs, missingInstallableIDs...)
-	updatesCnt.WithLabelValues("installable").Add(float64(len(installableIDs)))
-	utils.LogInfo("inventoryID", system.InventoryID, "installable", len(installableIDs), "installable advisories")
+	return extendedAdvisories, missingNames
+}
 
-	applicableIDs, missingApplicableNames, err := getAdvisoriesFromDB(tx, applicableNames)
+func loadMissingNamesIDs(missingNames []string, extendedAdvisories *extendedAdvisoryMap) error {
```
[the remark this answers is not in the document]
Comment (Dugowitch) [reply evaluator/evaluate_advisories.go:73]: You are right, it's unnecessary.

File: evaluator/evaluate_advisories.go [hunk 2]
```diff
@@ -12,162 +12,146 @@ import (
...
 	if err != nil {
 		return nil, err
 	}
 
-	advisoriesMap := make(map[string]models.SystemAdvisories, len(advisories))
-	for _, advisory := range advisories {
-		advisoriesMap[advisory.Advisory.Name] = advisory
-	}
-	return advisoriesMap, nil
+	parseStored(stored, reported, &extendedAdvisories)
```
[the remark this answers is not in the document]
Comment (Dugowitch) [reply evaluator/evaluate_advisories.go:120]: I guess both can be put into `evaluateChanges`. I just wanted to divide the logic into shorter functions, but it's not really needed.

File: evaluator/evaluate_advisories.go [hunk 3]
```diff
@@ -12,162 +12,146 @@ import (
...
-	nUnknown := len(installableNames) - len(installableIDs)
-	if nUnknown > 0 {
-		utils.LogInfo("inventoryID", system.InventoryID, "unknown", nUnknown, "unknown installable advisories")
-		updatesCnt.WithLabelValues("unknown").Add(float64(nUnknown))
-	}
-	missingInstallableIDs, err := storeMissingAdvisories(missingInstallableNames)
-	if err != nil {
-		return nil, nil, nil, errors.Wrap(err, "unable to store unknown installable advisories in db")
+// PasrseReported evaluates changes of type Add/Update/Keep and tracks them in extendedAdvisoryMap.
+func pasrseReported(stored SystemAdvisoryMap, reported map[string]int) (extendedAdvisoryMap, []string) {
```
[the remark this answers is not in the document]
Comment (Dugowitch) [reply evaluator/evaluate_advisories.go:43]: I will put it into `evaluateChanges` as well as `parseStored`.

File: evaluator/evaluate_advisories.go [hunk 4]
```diff
@@ -351,3 +368,17 @@ func updateAdvisoryAccountData(tx *gorm.DB, system *models.SystemPlatform, delet
...
+}
+
+type extendedAdvisoryMap map[string]extendedAdvisory
+
+const (
+	undefinedChangeType int = iota
+	enhancement
+	bugfix
+	security
+)
```
[the remark this answers is not in the document]
Comment (Dugowitch) [reply evaluator/evaluate_advisories.go:384]: 👍

**Comment by Dugowitch** [comment]
New PR was created. See #1461.

**Pull request closed without merging**